### ISSUE TITLE
Stereo added but no landmarks points and poses yet and stereo inertial mode is left

### ISSUE THREAD

**Issue by Shivam7Sharma** [suchetanrs/ORB-SLAM3-ROS2-Docker#22]
Hi @suchetanrs ,

I am not very familiar with github pull request. I forked this repo and created a new branch in that fork with all the changes. In this `stereo-working` branch, the map points are getting published. I am not sure about the odometry, but I can see the camera pose in Rviz in real time. I will later try to add stereo inertial. 

Also, is it possible to add landmark poses and points? I don't think it is currently a working feature.

The branch: https://github.com/Shivam7Sharma/ORB-SLAM3-ROS2-Docker.git

**Comment by suchetanrs** [comment]
Hi @Shivam7Sharma 
Thanks a lot for your new ticket.

Just to be sure so that I can recreate the issue you face on my system as well, you tried this with the Euroc dataset converted to a ros2 bag. Am I right?

Also, I'm not sure I fully understand what you mean by the landmark poses and points. Currently, whatever is being published as map points are the landmarks recorded. I have a further optimized method to publish the map points which I will push soon. This won't include all the map points from all the keyframes but only the ones being currently viewed.

Will have a look at your fork tomorrow. Thanks.

**Comment by Shivam7Sharma** [comment]
> Hi @Shivam7Sharma Thanks a lot for your new ticket.
> 
> Just to be sure so that I can recreate the issue you face on my system as well, you tried this with the Euroc dataset converted to a ros2 bag. Am I right?
> 
> Also, I'm not sure I fully understand what you mean by the landmark poses and points. Currently, whatever is being published as map points are the landmarks recorded. I have a further optimized method to publish the map points which I will push soon. This won't include all the map points from all the keyframes but only the ones being currently viewed.
> 
> Will have a look at your fork tomorrow. Thanks.

Yes, I use Euroc dataset MH01 easy dataset and converted to ROS 2 rosbag. I wanted to ask what the following publishers are publishing:
```
        visibleLandmarksPub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("visible_landmarks", 10);
        visibleLandmarksPose_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("visible_landmarks_pose", 10);
```

They are used in the services created in your code and one of them requires a request with a pose to give a response.

**Comment by Shivam7Sharma** [comment]
I have also noticed that the ground plane of point cloud is not parallel to the rviz ground plane. I have tested this with MH01 Easy dataset. Let me know if you observe the same thing.

**Comment by suchetanrs** [comment]
Hey @Shivam7Sharma I did not have a look at your stereo fork yet but quickly tried the RGBD mode to reproduce your TF errors. I was able to reproduce it and I suppose the errors are expected.

Just to clarify:
This wrapper does not publish an odometry on a topic. It simply publishes the required TF. 
```no_odometry_mode``` is used when the robot you are running the SLAM on does not have any on-board odometry and the wrapper must publish the TF between map and base_link directly.
If your robot does have an on-board odometry module (IMU, Wheel odometry etc) that publish the odometry onto a topic, the wrapper can use this to publish the transform between map and odom instead of map and base_link instead. This is done to follow the convention used similar to other SLAM systems to avoid jumps. 
This parameter has nothing to do with the SLAM publishing an odometry. It will publish the TF regardless of this param's value.
However, you received the TF errors because you set this param's value to false but there was no odometry published for the wrapper to use. If you don't have an odometry module, set it to true to avoid issues. Hope this clarifies :p

**Comment by suchetanrs** [comment]
For your second question, 
```
        visibleLandmarksPub_ = this->create_publisher<sensor_msgs::msg::PointCloud2>("visible_landmarks", 10);
        visibleLandmarksPose_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("visible_landmarks_pose", 10);
```
I added this primarily to visualize the response of the service call you mentioned.
This service is used to get the map points / landmarks that could be visible from any arbitrary pose that will be given in the request of the service. The response will contain the landmarks. The above two publishers are used to visualize the response.
> It would be cool to have all the keyframes and currently viewed points in RVIZ as separate topics. Is it possible to do that?

The keyframes are published already at regular intervals https://github.com/suchetanrs/ORB-SLAM3-ROS2-Docker/blob/eaa8db697c0665e740e9425838d1e284783ee3f4/orb_slam3_ros2_wrapper/src/rgbd/rgbd-slam-node.cpp#L191
However, they cannot be seen on Rviz since it's a custom message type. Yes, it would be cool to publish them as a pose array (parametrized if the user needs it to be published).

As mentioned in an earlier issue https://github.com/suchetanrs/ORB-SLAM3-ROS2-Docker/issues/18 , publishing all the landmarks at once seriously slows down the tracking. I believe what you said can be done. However, it might need some work. It would essentially be calling the function used in the service we talked above with the latest tracked pose. However, if you are open to working on this, let's discuss this further with more thought.
